book/src/rfc/const_generic_fn.md

```diff
@@ -0,0 +1,224 @@
+    /// Structures behind an array can be recursive, so it embeds the ComputExpr.
+    Array(Vec<ComputedExpr<F, V>>),
+    Bool(Var<F, V>),
+    Field(Var<F, V>),
+    /// Access to a variable in the scope.
+    Access(Access<F, V>),
+    /// Represents the results of a builtin function call.
+    /// Because we don't know the exact type of the result, we store it as a Var.
+    /// We may deprecate this once it is able to type check the builtin functions,
+    /// so that the result can be inferred.
+    FnCallResult(Var<F, V>),
+}
+```
+
+### Assignment using `ComputedExpr`
+
+Instead of using `VarOrRef` to narrow down and do the updating of `cvars`, we  use the `ComputedExpr` to represent the access path, which is iteratively determined when stepping through the computations of `Expr::FieldAccess` and `Expr::ArrayAccess`.
+
+When it comes to the computation for `Expr::Assignment`, it can use the access path embedded in a `ComputedExprKind::Access` to locate a nested `ComputedExpr` to swap with a new one. Then it saves the `VarInfo` with same name `houses` as local variable but holding the updated tree of `ComputedExpr` that represents a new set of `cvars`.
+
+### Examples
+The following are examples of how const generic can be used in functions, once this RFC is implemented.
+
+#### Declare a const argument
+Enable functions to define a constant argument, which can be used to define arrays with a default element or as an array size of the return type:
+
+```rust
+fn const_generic(const cst: Field) -> [Field; cst] {
+    let mut xx = [1; cst];
+    for ii in 1..cst {
+        xx[ii] = 2;
+    }
+    return xx;
+}
+```
+
+#### Builtin functions
+
+Allow builtin functions to be defined with const arguments:
+
+```rust
+use std::to_bits;
+use std::from_bits;
+
+const num_bits = 8;
+
+fn main() {
+    let val1 = 101;
+    let bits = to_bits(num_bits, val); // `bits` is an array of length `num_bits`
+    let val2 = from_bits(bits); // `from_bits` builtin impl knows the length of the array
+    assert_eq(val1, val2);
+}
+```
+
+The `to_bits` builtin function accepts a const variable to determine the size of the return array.
+
+```rust
+fn to_bits(const num_bits: Field, val: Field) -> [Field; num_bits]
+```
+
+The `from_bits` builtin function can determine the number of bits to convert based on the information that the `ComputedExprKind::Array` embeds in the `bits` variable. 
+
+#### Use ComputedExpr in builtin functions
+
+Psuedo code to show how the bit builtin functions could be implemented with the `ComputedExpr`:
+
+```rust
+fn to_bits<B: Backend>(
+    compiler: &mut CircuitWriter<B>,
+    vars: &[VarInfo<B::Field, B::Var>],
+    span: Span,
+) -> Result<Option<ComputedExpr<B::Field, B::Var>>> {
+    // should be two input vars
+    assert_eq!(vars.len(), 2);
+
+    // first var is the number of bits
+    let num_var = &vars[0];
+    assert!(matches!(num_var.expr.kind, ComputedExprKind::Field));
```

Comment:
Yeah, these checks should be already covered by the type checker. They are probably unnecessary here.

On the related issue, I think we have to check the const generic var is consistent with the actual array size resulted from the function call:

```rust
fn to_bits(const num_bits: Field, val: Field) -> [Field; num_bits]
```

Because the type checker doesn't know the actual value of `num_bits`, we would need to check the provided const generic argument against the size of the resulted array.